### PR TITLE
Feature/fork directive

### PR DESCRIPTION
**Pull request by csadorf**
#  Description

Implement the fork-directive that allows users to explicitly trigger forking of operations.

## Motivation and Context

Some operations that are parallelized should not be run within the interpreter process and should always be run in a sub-process. There is no explicit way of enforcing forking of Python-function-based operation at the moment.

There has been some discussion in the past about whether this should be better be a separate decorator since directives have previously been used primarily in the context of templates, however I believe that there are very strong arguments for making this a directive:

 * It is extremely easy to implement this behavior as a directive. The decorator is supposed to behave like a directive and for example is automatically an optional function of job, no extra handling needed.
 * We avoid needing to carry another parameter through all the various instances of flow-operations, job-operations, etc. That is especially important for the pending groups-feature, where would need to track forking as well.
* Directives are a powerful tool to control execution and submission behavior on a per-operation basis. Let's expand  on that.
* We avoid cluttering the package root namespace. We can always make this a top-level decorator at some point if we are so inclined.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

File: flow/operations.py [hunk 1]
```diff
@@ -104,6 +104,10 @@ class directives(object):
 
     Directives can for example be used to provide information about required resources
     such as the number of processes required for execution of parallelized operations.
+
+    In addition, you can use the `@directives(fork=True)` directive to enforce that a
+    particular  operation is always executed within a subprocess and not within the
```
Comment (bdice) [inline flow/operations.py]:
```suggestion
    particular operation is always executed within a subprocess and not within the
```

File: flow/operations.py [hunk 2]
```diff
@@ -104,6 +104,10 @@ class directives(object):
 
     Directives can for example be used to provide information about required resources
     such as the number of processes required for execution of parallelized operations.
+
+    In addition, you can use the `@directives(fork=True)` directive to enforce that a
```
Comment (bdice) [inline flow/operations.py]:
```suggestion
    In addition, you can use the `@directives(fork=True)` decorator to enforce that a
```

File: tests/define_test_project.py [hunk 3]
```diff
@@ -38,10 +39,15 @@ def op1(job):
...
+def _need_to_fork(job):
+    return job.doc.get('fork')
+
+
 @TestProject.operation
+@flow.directives(fork=_need_to_fork)
 @TestProject.post.true('test')
 def op2(job):
-    job.document.test = True
+    job.document.test = os.getpid()
```
Comment (bdice) [inline tests/define_test_project.py]: Good idea for the test. 👍

File: flow/operations.py (hunk 2 above)
Comment (csadorf) [reply flow/operations.py]: The thing I'm referencing is a decorator. Too confusing?

File: flow/operations.py [hunk 4]
```diff
@@ -104,6 +104,10 @@ class directives(object):
...
+    particular operation is always executed within a subprocess and not within the
+    Python interpreter's process even if there are no other reasons that would prevent that.
     """
```
Comment (csadorf) [inline flow/operations.py]:
```suggestion
+    .. note::
+
+        Setting `fork=False` will not prevent forking if there are other reasons for forking,
+        such as a deviating executable or a timeout argument.
    """
```

File: flow/operations.py (hunk 2 above)
Comment (bdice) [reply flow/operations.py]: I think the word directive is directly implied by `@directives` but "decorator" might be more helpful for new users. Accept or reject the suggestion at your discretion.

File: flow/operations.py (hunk 4 above)
Comment (bdice) [reply flow/operations.py]: "deviating executable" is not a word I recognize and couldn't find on Google. Maybe there's a better word? "An executable that creates a fork on its own"?

File: flow/operations.py (hunk 2 above)
Comment (csadorf) [reply flow/operations.py]: I'd prefer to leave as-is.

File: flow/operations.py (hunk 4 above)
Comment (csadorf) [reply flow/operations.py]: I've updated my comment, better?

**Comment by bdice** [comment]
@csadorf Can you comment on whether this solves issue #115, and how it relates to the closed PR #132? The code looks substantially different.

**Comment by csadorf** [comment]
> @csadorf Can you comment on whether this solves issue #115, and how it relates to the closed PR #132? The code looks substantially different.

I decided to create a new PR instead of re-opening #132 .

The issue #115 would be addressed with this PR, albeit it would require users to use `@directives(fork=True)` instead of `@fork`.

File: flow/operations.py (hunk 4 above)
Comment (csadorf) [reply flow/operations.py]: fixed typo

Comment (csadorf) [reply flow/operations.py]: Using timeout as example instead of executable.

File: flow/project.py [hunk 5]
```diff
@@ -1581,21 +1581,36 @@ def _run_operations_in_parallel(self, pool, pickle, operations, progress, timeou
...
-            self._operation_functions[operation.name](operation.job)
-        else:   # need to fork
+        # Check if we need to fork for operation execution...
+        if (
+            # The 'fork' directive was provided and evaluates to True:
+            operation.directives.get('fork', False)
+            # Separate process needed to cancel with timeout:
+            or timeout is not None
+            # The operation function is not registered with the class:
+            or operation.name not in self._operation_functions
```
Comment (bdice) [inline flow/project.py]: @csadorf Can you help me understand how this condition could happen? Is this from some kind of legacy `operations.py` syntax?

Comment (csadorf) [reply flow/project.py]: You could register an operation with `add_operation`. Operations don't *have* to be tied to functions.

File: flow/project.py [hunk 6]
```diff
@@ -1581,21 +1581,36 @@ def _run_operations_in_parallel(self, pool, pickle, operations, progress, timeou
...
+            # The specified executable is not the same as the interpreter instance:
+            or operation.directives.get('executable', sys.executable) != sys.executable
+        ):
+            # ... need to fork:
+            logger.debug(
+                "Forking to execute operation '{}' with "
+                "cmd '{}'.".format(operation, operation.cmd))
```
Comment (mikemhenry) [inline flow/project.py]: Would it be worth printing the pid here?

Comment (csadorf) [reply flow/project.py]: The PID is not known at this point. I'm not sure how difficult it would be to determine the PID.

**Pull request merged**